Error: sendAsync not overwritten

Continuation of #1940, which was really a different bug.
Since the fix to that issue was published, some users are getting an error when trying to interact with the MetaMask Web3, with an error of:
```
core.es5.js:1020
ERROR Error: Uncaught (in promise): Error: MetamaskInpageProvider - sendAsync not overwritten
Error: MetamaskInpageProvider - sendAsync not overwritten
    at MetamaskInpageProvider.sendAsync (inpage.js:254)
    at EthRPC.sendAsync (index.js:48)
    at index.js:96
    at new ZoneAwarePromise (zone.js:861)
    at Eth.outputMethod [as sendTransaction] (index.js:53)
    at Object.newContract [as new] (index.js:151)
    at DeployContractsComponent.deployWill (deploy-contracts.component.ts:124)
    at Object.eval [as handleEvent] (DeployContractsComponent.html:22)
    at handleEvent (core.es5.js:12014)
    at callWithDebugContext (core.es5.js:13475)
    at Object.debugHandleEvent [as handleEvent] (core.es5.js:13063)
    at dispatchEvent (core.es5.js:8607)
    at core.es5.js:9218
    at HTMLButtonElement.<anonymous> (platform-browser.es5.js:2651)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:425)
    at Object.onInvokeTask (core.es5.js:3881)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:424)
    at Zone.webpackJsonp.../../../../zone.js/dist/zone.js.Zone.runTask (zone.js:192)
    at ZoneTask.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneTask.invokeTask [as invoke] (zone.js:499)
    at invokeTask (zone.js:1427)
    at HTMLButtonElement.globalZoneAwareCallback (zone.js:1445)
    at MetamaskInpageProvider.sendAsync (inpage.js:254)
    at EthRPC.sendAsync (index.js:48)
    at index.js:96
    at new ZoneAwarePromise (zone.js:861)
    at Eth.outputMethod [as sendTransaction] (index.js:53)
    at Object.newContract [as new] (index.js:151)
    at DeployContractsComponent.deployWill (deploy-contracts.component.ts:124)
    at Object.eval [as handleEvent] (DeployContractsComponent.html:22)
    at handleEvent (core.es5.js:12014)
    at callWithDebugContext (core.es5.js:13475)
    at Object.debugHandleEvent [as handleEvent] (core.es5.js:13063)
    at dispatchEvent (core.es5.js:8607)
    at core.es5.js:9218
    at HTMLButtonElement.<anonymous> (platform-browser.es5.js:2651)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:425)
    at Object.onInvokeTask (core.es5.js:3881)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:424)
    at Zone.webpackJsonp.../../../../zone.js/dist/zone.js.Zone.runTask (zone.js:192)
    at 
```

Pinging users who experienced it, @andytudhope and @JamesKnippel. Can you each describe your dev stacks in detail? Are you both using React? Babel? Looking for tools that might be conflicting with MetaMask.

```
vendor.bundle.js:194751 ERROR Error: Uncaught (in promise): Error: MetamaskInpageProvider - sendAsync not overwritten
Error: MetamaskInpageProvider - sendAsync not overwritten
    at MetamaskInpageProvider.sendAsync (inpage.js:254)
    at EthRPC.sendAsync (index.js:48)
    at index.js:96
    at new ZoneAwarePromise (polyfills.bundle.js:3406)
    at Eth.outputMethod [as sendTransaction] (index.js:53)
    at Object.newContract [as new] (index.js:151)
    at DeployContractsComponent.deployWill (main.bundle.js:190)
    at Object.eval [as handleEvent] (DeployContractsComponent.ngfactory.js:173)
    at handleEvent (core.es5.js:12014)
    at callWithDebugContext (core.es5.js:13475)
    at Object.debugHandleEvent [as handleEvent] (core.es5.js:13063)
    at dispatchEvent (core.es5.js:8607)
    at core.es5.js:9218
    at HTMLButtonElement.<anonymous> (platform-browser.es5.js:2651)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2970)
    at Object.onInvokeTask (core.es5.js:3881)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2969)
    at Zone.webpackJsonp.../../../../zone.js/dist/zone.js.Zone.runTask (polyfills.bundle.js:2737)
    at ZoneTask.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneTask.invokeTask [as invoke] (polyfills.bundle.js:3044)
    at invokeTask (polyfills.bundle.js:3972)
    at HTMLButtonElement.globalZoneAwareCallback (polyfills.bundle.js:3990)
    at MetamaskInpageProvider.sendAsync (inpage.js:254)
    at EthRPC.sendAsync (index.js:48)
    at index.js:96
    at new ZoneAwarePromise (polyfills.bundle.js:3406)
    at Eth.outputMethod [as sendTransaction] (index.js:53)
    at Object.newContract [as new] (index.js:151)
    at DeployContractsComponent.deployWill (main.bundle.js:190)
    at Object.eval [as handleEvent] (DeployContractsComponent.ngfactory.js:173)
    at handleEvent (core.es5.js:12014)
    at callWithDebugContext (core.es5.js:13475)
    at Object.debugHandleEvent [as handleEvent] (core.es5.js:13063)
    at dispatchEvent (core.es5.js:8607)
    at core.es5.js:9218
    at HTMLButtonElement.<anonymous> (platform-browser.es5.js:2651)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2970)
    at Object.onInvokeTask (core.es5.js:3881)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2969)
    at Zone.webpackJsonp.../../../../zone.js/dist/zone.js.Zone.runTask (polyfills.bundle.js:2737)
    at ZoneTask.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneTask.invokeTask [as invoke] (polyfills.bundle.js:3044)
    at invokeTask (polyfills.bundle.js:3972)
    at HTMLButtonElement.globalZoneAwareCallback (polyfills.bundle.js:3990)
    at resolvePromise (polyfills.bundle.js:3340)
    at new ZoneAwarePromise (polyfills.bundle.js:3409)
    at Eth.outputMethod [as sendTransaction] (index.js:53)
    at Object.newContract [as new] (index.js:151)
    at DeployContractsComponent.deployWill (main.bundle.js:190)
    at Object.eval [as handleEvent] (DeployContractsComponent.ngfactory.js:173)
    at handleEvent (core.es5.js:12014)
    at callWithDebugContext (core.es5.js:13475)
    at Object.debugHandleEvent [as handleEvent] (core.es5.js:13063)
    at dispatchEvent (core.es5.js:8607)
    at core.es5.js:9218
    at HTMLButtonElement.<anonymous> (platform-browser.es5.js:2651)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2970)
    at Object.onInvokeTask (core.es5.js:3881)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2969)
    at Zone.webpackJsonp.../../../../zone.js/dist/zone.js.Zone.runTask (polyfills.bundle.js:2737)
    at ZoneTask.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneTask.invokeTask [as invoke] (polyfills.bundle.js:3044)
    at invokeTask (polyfills.bundle.js:3972)
    at HTMLButtonElement.globalZoneAwareCallback (polyfills.bundle.js:3990)

```
This is the error when I try to deploy using ethjs, I posted my deployment strategy in the last thread I believe. 

this is web3's when i change branches back over to my web3 branch coming from ethjs, 
mind you, my working strategy used web3 and truffle-contract module but would still provide this error very occasionally 
```
Error: MetamaskInpageProvider - sendAsync not overwritten
    at MetamaskInpageProvider.sendAsync (inpage.js:254)
    at Provider.webpackJsonp.../../../../truffle-contract/contract.js.Provider.sendAsync (contract.js:23)
    at RequestManager.webpackJsonp.../../../../truffle-contract/node_modules/web3/lib/web3/requestmanager.js.RequestManager.sendAsync (requestmanager.js:80)
    at Object.get [as getNetwork] (property.js:116)
    at contract.js:477
    at new ZoneAwarePromise (polyfills.bundle.js:3406)
    at Function.detectNetwork (contract.js:468)
    at Function.new (contract.js:327)
    at DeployContractsComponent.deployWill (main.bundle.js:204)
    at Object.eval [as handleEvent] (DeployContractsComponent.ngfactory.js:173)
    at handleEvent (core.es5.js:12014)
    at callWithDebugContext (core.es5.js:13475)
    at Object.debugHandleEvent [as handleEvent] (core.es5.js:13063)
    at dispatchEvent (core.es5.js:8607)
    at core.es5.js:9218
    at HTMLButtonElement.<anonymous> (platform-browser.es5.js:2651)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2970)
    at Object.onInvokeTask (core.es5.js:3881)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2969)
    at Zone.webpackJsonp.../../../../zone.js/dist/zone.js.Zone.runTask (polyfills.bundle.js:2737)
    at ZoneTask.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneTask.invokeTask [as invoke] (polyfills.bundle.js:3044)
    at invokeTask (polyfills.bundle.js:3972)
deploy-contracts.component.ts:150 Error: MetamaskInpageProvider - sendAsync not overwritten
    at MetamaskInpageProvider.sendAsync (inpage.js:254)
    at Provider.webpackJsonp.../../../../truffle-contract/contract.js.Provider.sendAsync (contract.js:23)
    at RequestManager.webpackJsonp.../../../../truffle-contract/node_modules/web3/lib/web3/requestmanager.js.RequestManager.sendAsync (requestmanager.js:80)
    at Object.get [as getNetwork] (property.js:116)
    at contract.js:477
    at new ZoneAwarePromise (polyfills.bundle.js:3406)
    at Function.detectNetwork (contract.js:468)
    at Function.new (contract.js:327)
    at DeployContractsComponent.deployWill (main.bundle.js:204)
    at Object.eval [as handleEvent] (DeployContractsComponent.ngfactory.js:173)
    at handleEvent (core.es5.js:12014)
    at callWithDebugContext (core.es5.js:13475)
    at Object.debugHandleEvent [as handleEvent] (core.es5.js:13063)
    at dispatchEvent (core.es5.js:8607)
    at core.es5.js:9218
    at HTMLButtonElement.<anonymous> (platform-browser.es5.js:2651)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2970)
    at Object.onInvokeTask (core.es5.js:3881)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2969)
    at Zone.webpackJsonp.../../../../zone.js/dist/zone.js.Zone.runTask (polyfills.bundle.js:2737)
    at ZoneTask.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneTask.invokeTask [as invoke] (polyfills.bundle.js:3044)
    at invokeTask (polyfills.bundle.js:3972)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,10 @@
 
 ## Current Master
 
+## 3.9.13 2017-9-8
+
+- Changed the way we initialize the inpage provider to fix a bug affecting some developers.
+
 ## 3.9.12 2017-9-6
 
 - Fix bug that prevented Web3 1.0 compatibility

--- a/app/manifest.json
+++ b/app/manifest.json
@@ -1,7 +1,7 @@
 {
   "name": "MetaMask",
   "short_name": "Metamask",
-  "version": "3.9.12",
+  "version": "3.9.13",
   "manifest_version": 2,
   "author": "https://metamask.io",
   "description": "Ethereum Browser Extension",

--- a/app/scripts/lib/inpage-provider.js
+++ b/app/scripts/lib/inpage-provider.js
@@ -40,30 +40,36 @@ function MetamaskInpageProvider (connectionStream) {
   // start and stop polling to unblock first block lock
 
   self.idMap = {}
-  // handle sendAsync requests via asyncProvider
-  self.sendAsync = function (payload, cb) {
-    // rewrite request ids
-    var request = eachJsonMessage(payload, (_message) => {
-      const message = Object.assign({}, _message)
-      const newId = createRandomId()
-      self.idMap[newId] = message.id
-      message.id = newId
+}
+
+// handle sendAsync requests via asyncProvider
+// also remap ids inbound and outbound
+MetamaskInpageProvider.prototype.sendAsync = function (payload, cb) {
+  const self = this
+
+  // rewrite request ids
+  const request = eachJsonMessage(payload, (_message) => {
+    const message = Object.assign({}, _message)
+    const newId = createRandomId()
+    self.idMap[newId] = message.id
+    message.id = newId
+    return message
+  })
+
+  // forward to asyncProvider
+  self.asyncProvider.sendAsync(request, (err, _res) => {
+    if (err) return cb(err)
+    // transform messages to original ids
+    const res = eachJsonMessage(_res, (message) => {
+      const oldId = self.idMap[message.id]
+      delete self.idMap[message.id]
+      message.id = oldId
       return message
     })
-    // forward to asyncProvider
-    asyncProvider.sendAsync(request, function (err, res) {
-      if (err) return cb(err)
-      // transform messages to original ids
-      eachJsonMessage(res, (message) => {
-        var oldId = self.idMap[message.id]
-        delete self.idMap[message.id]
-        message.id = oldId
-        return message
-      })
-      cb(null, res)
-    })
-  }
+    cb(null, res)
+  })
 }
+
 
 MetamaskInpageProvider.prototype.send = function (payload) {
   const self = this
@@ -108,10 +114,6 @@ MetamaskInpageProvider.prototype.send = function (payload) {
     jsonrpc: payload.jsonrpc,
     result: result,
   }
-}
-
-MetamaskInpageProvider.prototype.sendAsync = function () {
-  throw new Error('MetamaskInpageProvider - sendAsync not overwritten')
 }
 
 MetamaskInpageProvider.prototype.isConnected = function () {

